Refactor Bar Series (#741)

Jonarw opened the pull request:
Fixes #741 and #729.

### Checklist

- [x] I have included examples or tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file
- [x] I have cleaned up the commit history (use rebase and squash)

### Changes proposed in this pull request:

- Extract functionality related to `BarSeries` from `CategoryAxis` into new `BarSeriesManager` class
- Refactor bar series:
    - Change `CategorizedSeries` (a relict from the time when we had a separate `ColumnSeries`) to `BarSeriesBase<T>`
    - Move a lot of duplicated code from `BarSeries`, `TornadoBarSeries` and `IntervalBarSeries` into `BarSeriesBase<T>`
- My goal was to make the `BarSeriesManager` completely transparent to the `PlotModel`. This led to the somewhat strange `CheckIn...()` methods, which `BarSeriesBase` calls during updating and rendering. I could not think of a better solution for this, if someone else has one, please share :)
- I removed `Axis.UpdateFromSeries(...)` and `Series.UpdateValidData()`, as these seemed to be there just for the bar series and are not needed anymore


@oxyplot/admins

VisualMelon commented:
There seems to be a fairly serious regression: when there are disabled series (e.g. checking them off the legend in "All in one"), the plot goes completely insane when you pan/zoom/do anything.

Jonarw commented:
Yeah... thanks for noticing!
This in the end is a symptom of the weird interaction between series and manager... I guess I'll have to come up with a better way of doing this

Jonarw commented:
The latest commit fixes the above mentioned bug.

Outstanding concerns:

- If a bar series never has been updated, it will not render until the plot is updated with `updateData==true`. This occurs for example when enabling a bar series that was never visible by clicking on its legend entry (e.g. Series 3 in the All-in-One example). 
However this is consistent with the behavior before this PR and not necessarily something that is the fault of the series itself.
- I still don't really like the way the `BarSeriesBase` and the `BarSeriesManager` interact. However I can't think of a better alternative and I think that's the best we can do without involving the `PlotModel`, which was one of my goals with this.

VisualMelon commented:
It still seems to be possible to confuse stacked bar graphs by turning series visibility on and off, though panning/zooming seems to fix it.

![image](https://user-images.githubusercontent.com/5646475/82749114-e7f1d280-9d9e-11ea-9519-29766c7d66b1.png)

It also seems that if a series 'starts' invisible (e.g. series 3 in `AllInOne`) that it doesn't display when you check it on or off on the legend. This isn't changed by zooming/panning.

Jonarw commented:
Yes, both of these issues in the end come down to the fact that the bar offsets are determined by `BarSeriesManager` during plot update. When you change the visibility of a graph without updating the plot (as is done when you click on the legend entry), it can become confused.

VisualMelon commented:
Right. I won't open that issue immediately, because it does seem to be specific to `BarSeries`, and I'd like to properly review your code before diving into the old stuff on my own.

Jonarw commented:
(rebased)
I took another look at the code in this PR and I am not really sure what to make of it.
Sure, the current design where the `CategoryAxis` is responsible for telling the `BarSeries` where to render is 'not ideal', and this PR definitely improves on that with the introduction of the `BarSeriesManager` class. But I still absolutely don't like how `BarSeriesManager`s are created by `BarSeriesBase` itself (who creates their own manager??) and how the `BarSeriesManager` kind of has to guess when all of its series have been updated / rendered.

This is kind of a consequence of my goal to make the `BarSeriesManager` completely transparent to the `PlotModel`. I think if we want to keep it that way, this is pretty much the best we can do. But maybe this goal is flawed to start with?

So I think we have four possibilities:
1. Merge this PR like it is (after some more testing and bug fixing), it works well enough for the moment
2. Change it such that the `BarSeriesManager` is managed by the `PlotModel`. `IPlotModel.Update()` would maybe call a dedicated method `UpdateBarSeries()`, where all managers are set up and updated. Similarly, `IPlotModel.Render()` would call a dedicated method `RenderBarSeries()`.
3. Solve this problem some other way I can't think of
4. Forget this PR ever existed, maybe someone else can take a shot at it in the future and figure out a better design

Input is appreciated :)

VisualMelon commented:
@Jonarw sorry, completely forgot to look at this.

The bit that looks most dodgy to me is the observation that the manager has to guess at when to do its own updates: this could be solved by adding it to a list of things that PlotModel 'updates' once it has done everything else. Another set that are updated before everything else would help with things like slaving axes.

This could just be done with events (see. https://github.com/oxyplot/oxyplot/issues/388#issuecomment-75500509).

Or have I missed the point?

Jonarw commented:
I have also thought about triggering updates of the manager by events.
The problem that remains in that case: When is the manager created in the first place? And when is it subscribed to the relevant PlotModel event? 
At the current status of this PR, the manager is created by the BarSeries itself (by the first one that gets updated), but I guess we would also want to avoid this... So we'd have to create it somewhere in the `PlotModel` and we are back to what I would have liked to avoid.

VisualMelon commented:
I don't think it's a problem for a `BarSeries` to create its manager. One could argue that's the only way that makes any sense, because nobody else necessarily knows when a `BarSeries` is added. A bigger issue in my eyes is that the `BoxSeries` have to keep track of its manager. It also feels like `BarSeriesManager` has too much work in not only managing the bar series but deciding which bar series to manage and when to perform these managerial tasks.

What is the issue with having `CategoryAxis` do this? (i.e. it's good to have `BarSeriesManager`, but could it be held and invoked by `CategoryAxis`?). If not the axes themselves, then I think there needs to be a single objects per `PlotModel` to manage the `BarSeriesManager` instances (which could be renamed to `BarSeriesArranger`, or w/e), because anything else will create a lot of additioanl complexity (complexity which current exists in `BarSeriesManger` but will be hard to test and I don't trust). [Original draft with some additional musings/insanity](https://gist.github.com/VisualMelon/6849a2ac53ec6c603274b193c576f42a).

Jonarw commented:
Thanks for your input! It really helps to hear another perspective!

> What is the issue with having `CategoryAxis` do this?

I would have two issues with this:

- `CategoryAxis` should not assume it is only ever used for `BarSeries`, there could also be other types of series that use a `CategoryAxis` (in fact at work I use one of these). So I don't think there should be any logic related to `BarSeries` in `CategoryAxis`.
- There can be multiple `BarSeriesManager`s with the same `CategoryAxis` (where multiple value axes are present)

As a concept, I changed the implementation such that the `PlotModel` now manages the `BarSeriesManager`s. This for sure is not a perfect solution, but more a pragmatic one. At the very least it's better than what I had before and has the nice side effect that it seems to handle invisible series fairly well (though it's not fully tested yet).

If we encounter other similar issues in the future we could still add something like the `IService` concept @VisualMelon proposed, but for just this one application this might be overkill?

VisualMelon commented:
>  There can be multiple BarSeriesManagers with the same `CategoryAxis`

Good point. I'd completely overlooked that possibility.

> but for just this one application this might be overkill

Agreed. We can work out a suitable abstraction later when we have more than one example from which to work (though I was wondering if we could abstract out the layout code: it might make `SubPlots` and e.g. axis slaving easier to bodge without modifying the library).

Jonarw commented:
Did some more testing, added an example with non-stacked `BarSeries` and multiple value axes and fixed an issue with the Transpose() function, which did not work properly for series that are invisible initially (such as in the All-in-one example).

Everything seems to work well enough to me now. Fancy taking another look @VisualMelon ?

VisualMelon commented on Source/Examples/ExampleLibrary/Utilities/PlotModelUtilities.cs:
```diff
@@ -301,24 +301,38 @@ public static PlotModel Transpose(this PlotModel model)
+                if (series.XAxisKey == null)
                 {
-                    if (series.XAxis.Key == null)
+                    if (series.XAxis == null) // this can happen if the series in invisible initially
```
Typo: `is`

VisualMelon commented:
Nothing jumps out at me, and I'll try to take a proper look in the next couple of days.

VisualMelon commented:
Could do with an example with more than one CategoryAxis, e.g. https://gist.github.com/VisualMelon/a31c6a4f8d4388548f357d51b0615bd0

I can't find any unexpected behaviour, but I've not looked at the code yet.

VisualMelon commented on Source/Examples/ExampleLibrary/Utilities/PlotModelUtilities.cs:
```diff
@@ -301,24 +301,38 @@ public static PlotModel Transpose(this PlotModel model)
             foreach (var series in model.Series.OfType<XYAxisSeries>())
             {
-                if (series.XAxis != null && series.XAxisKey == null)
+                if (series.XAxisKey == null)
```
Would it be possible to simplify this by calling `EnsureAxes` on the series before otherwise interrogating it?

Jonarw replied:
I like the idea in general, but this is not available here to use, as it is `protected internal`...

VisualMelon replied:
Yep, sorry.

VisualMelon commented on Source/OxyPlot/Series/BarSeries/ErrorBarSeries.cs:
```diff
@@ -53,7 +53,7 @@ protected internal override void UpdateMaxMin()
             base.UpdateMaxMin();
 
             //// Todo: refactor (lots of duplicate code here)
```
Is this still needed?

Jonarw replied:
I would argue yes. The `UpdateMaxMin()` implementation of `ErrorBarSeries` is quite identical to the implementation of `BarSeries`, which I think this comment refers to.

VisualMelon commented on Source/OxyPlot/Series/BarSeries/IBarSeries.cs:
```diff
@@ -0,0 +1,73 @@
+        /// </summary>
+        void UpdateValidData();
+
+        // TODO: replace the following members by IReadOnlyList<BarItemBase> once we get rid of .NET 4.0.
```
There was a specific request to keep Net Framework 4.0 support quite recently (or atleast it was readded in #1452), so I'm not sure if these comments are useful.

Would it be insane to provide our own type to expose these members so that we don't have the interface splatted into other types', making a future refit painless for everyone (e.g. just change `CustomReadOnlyListWrapperType` with `IReadOnlyList` and update the writes: minimal code change for consumers). This type could even implement `IReadOnlyList` for compilation targets that support it, keeping conditions in one place. Are there other places we could use such a type, or is it just a barmy idea?

Jonarw replied:
_(I do not quite agree on supporting .NET 4.0 for much longer and I hate that it forces us to do these kinds of considerations which should be a non-issue nowadays, but that's not the issue here.)_

I think there are quite some places where `IList<T>` is used that should use `IReadOnlyList<T>` instead. In this specific case this was not possible because of lack of covariance.

Possibly slightly insane idea: How about we define `IReadOnlyList<T>` (and maybe also `IReadOnlyCollection<T>`) ourselves (conditionally compiled only for NET40) and a custom list class implementing this (again only for NET40)? 
I'll do some experiments to see if this could even remotely work.

VisualMelon replied:
I'd be hesitant to use the same names, because the 4.0 library could be used by 4.5 code and that might create confusion, but it would be nice to replace all the horridness with `IList<T>` with some well-typed horridness that can be quickly and easily removed if/when 4.0 support is dropped.

Jonarw replied:
Of course we could also use this [package](https://www.nuget.org/packages/ReadOnlyCollectionInterfaces/) ([GitHub link](https://github.com/mausch/ReadOnlyCollections)) for NET4.0. 

You are right, it would not be possible to use the 4.0 library from 4.5 (it would complain when loading the package about ambiguous types I think), but is this a problem?

VisualMelon replied:
Thanks for the link: I went looking for that project but couldn't find it. That said, I think it's probably too much for our purposes.

It shouldn't be ambiguous so long as the fully qualified name is distinct (which I think was the idea); I just think it could create confusion for users, and that being up-front that this is a thin wrapper over what is usually an array (I think) is a good idea. I suppose we need to decide whether it is important to improve the 4.0 interface, or just make life easier for everybody else. Something as simple as https://gist.github.com/VisualMelon/d8baadc9f28f8d5f1bf60bf9eaf8c318 might suffice.

Jonarw replied:
Well this package just contains the three interfaces `IReadOnlyCollection<T>`, `IReadOnlyList<T>` and `IReadOnlyDictionary<T>`, so I wouldn't necessarily say it's too much ;).
However I realized we can't use this anyway, since the assembly is not signed, so doesn't matter anyway. If we would want to go this way, we would have to define the interfaces ourselves (which wouldn't be too difficult).

Your solutions seems neat, especially the implicit cast from `List<T>`. However there is one thing I am a bit concerned about: When we ever finally get rid of .NET 4.0, this will be a breaking change for all interfaces using `ReadOnlyList<T>` (as they will be changed to `IReadOnlyList<T>`). Ok granted, probably this change will not be noticed by most consumers of the API, so maybe this is the best option we have...

VisualMelon replied:
Yeah, maybe I'm being silly, and we should just provide a conditionally defined `IReadOnlyList<T>` along with a wrapper type to re-expose `IList<T>`.

Jonarw replied:
I think this topic warrants its own issue, so I'll open a new one.

VisualMelon replied:
Good idea, and I don't think it should block this being merged.

Jonarw replied:
Well I don't mind leaving this open for a couple more days until we have figured out what to do rather than going back later and adapting this to whatever solution we settle on.

VisualMelon commented on Source/OxyPlot/Utilities/ArrayExtensions.cs:
```diff
@@ -9,11 +9,36 @@
+        /// </summary>
+        /// <param name="sequence">The sequence.</param>
+        /// <returns>The maximum value of the sequence, or 0 if the sequency is empty.</returns>
+        public static double MaxOrZero(this IEnumerable<double> sequence)
```
Seems a tad specific. If it took a default value as a parameter that would be nice (the deafult value is a concern for the calling domain), but I'm not sure it's worth the effortof coming up with a different name.

Jonarw replied:
Maybe we could just change it to `MaxOrDefault<T>(...)` and return `default(T)` if the sequence is empty?

VisualMelon replied:
The issue is that in either case the default value for e.g. doubles is meaningful (i.e. it's not something like `null` which explodes if you use it). I'd suggest that the default should be provided by the calling code, because only it can know if there is a meaningful default (if there isn't, it should be using `Max` and throwing or something like `TryGetMax`). I don't think it's a huge issue, but providing a default default (heh) of zero makes the method limited and easy to misuse with `structs`.

I don't think this is a big issue, so feel free to ignore it, but it feels odd for a `public` method.

Jonarw replied:
Replaced this by `MaxOrDefault` with a user-defined default value.
Also added a corresponding `MinOrDefault` because it would feel weird otherwise.

VisualMelon commented on Source/OxyPlot/Utilities/ArrayExtensions.cs:
```diff
@@ -9,11 +9,36 @@
+        /// <returns>The maximum value of the sequence, or 0 if the sequency is empty.</returns>
+        public static double MaxOrZero(this IEnumerable<double> sequence)
+        {
+            using var e = sequence.GetEnumerator();
```
`null` check would be nice on `sequence`

Jonarw replied:
Replaced this by `MaxOrDefault` with a user-defined default value.
Also added a corresponding `MinOrDefault` because it would feel weird otherwise.

VisualMelon commented:
There is a change of behaviour with regard to shared values axis: if 2 bar series share a category series but have separate value axes, then they would previously be grouped together. The new behaviour makes sense if the value axes are 'seperated' (as in the `MultipleValueAxes` example), but not if they 'overlap' (e.g. share a common zero position).

I don't think the old behaviour was really supportable or intentional given the the lack of support for axis slaving and the fact that the min/max go wrong if not manually specified - so I don't think this can be considered a regression -but I wonder if it would be a useful feature. The current system can't support it because it associates each manager with a single axis pair.

VisualMelon commented on Source/OxyPlot/Series/BarSeries/ErrorBarSeries.cs:
```diff
@@ -110,17 +110,8 @@ protected internal override void UpdateMaxMin()
             }
 
-            var valueAxis = this.GetValueAxis();
-            if (valueAxis.IsVertical())
```
Was this a hold-over from when Column/Bar series were separate concerns?

Jonarw replied:
Yes, exactly

VisualMelon commented on Source/OxyPlot/Series/BarSeries/BarSeries.cs:
```diff
@@ -682,11 +357,90 @@ protected Axis GetValueAxis()
+                var value = item.Value;
+
+                // Get base- and topValue
+                var baseValue = double.NaN;
+                if (this.IsStacked)
+                {
+                    baseValue = this.Manager.GetCurrentBaseValue(stackIndex, categoryIndex, value < 0);
+                }
+
+                if (double.IsNaN(baseValue))
+                {
+                    baseValue = this.BaseValue;
+                }
```
This isn't new in this PR, but it's a bit weird.

VisualMelon commented:
I've not looked at all the removed code, but all the new code looks good (and surprisingly simple) to me (apart from the minor concerns discussed above).

Jonarw commented:
(rebased on #1601)
As discussed this now shows #1601 'in action' and makes the `IBarSeries` interface much nicer.

Jonarw commented on Source/OxyPlot/Utilities/ArrayExtensions.cs:
```diff
@@ -9,11 +9,36 @@
+        /// <returns>The maximum value of the sequence, or 0 if the sequency is empty.</returns>
+        public static double MaxOrZero(this IEnumerable<double> sequence)
+        {
+            using var e = sequence.GetEnumerator();
```
Sorry, wanted to delete the accidentally misplaced comment, and for some reason it deleted your comment regarding the null check??

Anyway, I've added the null check.

VisualMelon commented:
Is there anything else you are planning to do with this or that you want another pair of eyes over?

Jonarw commented:
No, from my side I think this is pretty much done, except of course you (or anyone else) finds more issues.

(And by the way, I am not planning to squash the commits as I recently discovered GitHub's 'Squash and merge' feature. I think it is neat to preserve the full commit history on the original PR while still having a 'clean' history on the main branch.)

Jonarw commented:
Any objections to merging at this point?
@VisualMelon

VisualMelon commented:
None here.
